### PR TITLE
Refactor filter descriptor type checks in SwaggerGen

### PR DESCRIPTION
**Pull request by iskandersierra**
# Pull Request

Given that the public API allows to register a filter in this form without any specific validation on not assigning the Type property:

```csharp
genOptions.OperationFilterDescriptors.Add(new FilterDescriptor
{
    // Type = typeof(TdpHeaderParameter),
    FilterInstance = new TdpHeaderParameter(IRequestContext.UserIdKey, "The user ID."),
});
```

in which case, opening the swagger UI returns a `NullReferenceException` (in developer error page) with no reference to user source code.

Looking into the library source code, I found out that the "proper" way to do the registration is:

```csharp
genOptions.AddOperationFilterInstance(
    new TdpHeaderParameter(IRequestContext.TenantIdKey, "The tenant ID."));
```

but I just felt that the first option should also be available with minimal behavior changes, given that it is a possible choice in the public API.

## The issue or feature being addressed

> NOTE: I didn't create a specific issue

## Details on the issue fix or feature implementation

I added an alternative validation on the filter type for when the `FilterInstance` is set but not the `Type` of the `FilterDescriptor`.

### CODE REVIEW

File: src/Swashbuckle.AspNetCore.SwaggerGen/DependencyInjection/SwaggerGenOptionsExtensions.cs [hunk 1]
```diff
@@ -370,7 +370,6 @@ public static void AddSchemaFilterInstance<TFilter>(
             if (filterInstance == null) throw new ArgumentNullException(nameof(filterInstance));
             swaggerGenOptions.SchemaFilterDescriptors.Add(new FilterDescriptor
             {
-                Type = typeof(TFilter),
```
Comment (iskandersierra) [inline src/Swashbuckle.AspNetCore.SwaggerGen/DependencyInjection/SwaggerGenOptionsExtensions.cs:373]: There is not need to remove this assignments to property `Type`, even when this pass all the tests. Maybe someone out there is relying on the `Type` being assigned even when non needed by the library, because it is only used if the `FilterInstance` is not assigned.

I would revert the changes in this file.

Comment (martincostello) [reply src/Swashbuckle.AspNetCore.SwaggerGen/DependencyInjection/SwaggerGenOptionsExtensions.cs:373]: Please revert these changes - as you say, they are part of the visible external behaviour and someone might be using them.

File: src/Swashbuckle.AspNetCore.SwaggerGen/DependencyInjection/SwaggerGenOptions.cs [hunk 2]
```diff
@@ -32,5 +32,11 @@ public class FilterDescriptor
         public object[] Arguments { get; set; }
 
         public object FilterInstance { get; set; }
+
+        internal bool IsAssignableTo(Type type)
+        {
+            return FilterInstance != null && type.IsInstanceOfType(FilterInstance) ||
+                   Type != null && Type.IsAssignableTo(type);
```
Comment (martincostello) [inline src/Swashbuckle.AspNetCore.SwaggerGen/DependencyInjection/SwaggerGenOptions.cs:39]: Add parenthesis to make the order of operations obvious:
```suggestion
            return (FilterInstance != null && type.IsInstanceOfType(FilterInstance)) ||
                   (Type != null && Type.IsAssignableTo(type));
```

**Review by martincostello — COMMENTED** [review]
Can you add a test that validates the `NullReferenceException` has gone away with the change?

**Review by martincostello — APPROVED** [review]
